src/filesystem/sharing/shared_file.rs

```diff
@@ -0,0 +1,66 @@
+use base64::{engine::general_purpose::URL_SAFE, Engine as _};
+use serde::{Deserialize, Serialize};
+use wnfs::{common::dagcbor, libipld::Cid, private::share::SharePayload};
+
+use super::SharingError;
+
+#[derive(Debug, Clone)]
+pub struct SharedFile {
+    pub payload: SharePayload,
+    pub forest_cid: Cid,
+    pub file_name: String,
+    pub mime_type: String,
+    pub size: u64,
+}
+
+impl Serialize for SharedFile {
+    fn serialize<S>(&self, serializer: S) -> std::prelude::v1::Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let payload_bytes = dagcbor::encode(&self.payload).expect("failed to serialize payload");
```

Comment:
of course, the inbuilt WNFS serializer that runs via a `#[derive(..)]` is asymetric and cant reconstruct. so i'm getting around that by implementing a custom serialization on this type.